internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,321 @@
+) *DynamicScaleDownManager {
+
+	return &DynamicScaleDownManager{
+		BaseWorker: workers.BaseWorker{
+			Id:         uuid.New().String(),
+			WorkerType: dynamicScaleDownWorkerType,
+			Reconciler: reconciler,
+		},
+
+		dataplaneClusterConfig: dataplaneClusterConfig,
+		clusterProvidersConfig: clusterProvidersConfig,
+		kafkaConfig:            kafkaConfig,
+		clusterService:         clusterService,
+	}
+}
+
+func (m *DynamicScaleDownManager) Start() {
+	m.StartWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Stop() {
+	m.StopWorker(m)
+}
+
+func (m *DynamicScaleDownManager) Reconcile() []error {
+	var errList fleeterrors.ErrorList
+	if !m.dataplaneClusterConfig.IsDataPlaneAutoScalingEnabled() {
+		glog.Infoln("dynamic scaling is disabled. Dynamic scale down reconcile event skipped")
+		return nil
+	}
+	glog.Infoln("running dynamic scale down reconcile event")
+	defer m.logReconcileEventEnd()
+
+	err := m.processDynamicScaleDownReconcileEvent()
+	if err != nil {
+		errList.AddErrors(err)
+	}
+
+	return errList.ToErrorSlice()
+}
+
+func (m *DynamicScaleDownManager) processDynamicScaleDownReconcileEvent() error {
+	var errList fleeterrors.ErrorList
+	kafkaStreamingUnitCountPerClusterList, err := m.clusterService.FindStreamingUnitCountByClusterAndInstanceType()
+
+	if err != nil {
+		errList.AddErrors(err)
+		return errList
+	}
+
+	type processed struct {
+		indexesOfStreamingUnitForSameClusterID []int
+		processed                              bool
+	}
+
+	var alreadyProcessedClusters map[string]processed = map[string]processed{}
+
+	for i, kafkaStreamingUnitCountPerCluster := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := kafkaStreamingUnitCountPerCluster.ClusterId
+		existing, ok := alreadyProcessedClusters[clusterID]
+
+		if !ok {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: []int{i},
+				processed:                              false,
+			}
+		} else {
+			alreadyProcessedClusters[clusterID] = processed{
+				indexesOfStreamingUnitForSameClusterID: append(existing.indexesOfStreamingUnitForSameClusterID, i),
+				processed:                              existing.processed,
+			}
+		}
+	}
+
+	for _, suCount := range kafkaStreamingUnitCountPerClusterList {
+		clusterID := suCount.ClusterId
+		existing, alreadyProcessed := alreadyProcessedClusters[clusterID]
+		if alreadyProcessed && existing.processed { // skip already processed
+			continue
+		}
+
+		alreadyProcessedClusters[clusterID] = processed{
+			indexesOfStreamingUnitForSameClusterID: existing.indexesOfStreamingUnitForSameClusterID,
+			processed:                              true,
+		}
+
+		if suCount.Status != api.ClusterReady.String() { // do not consider non ready cluster for scale down evaluation
+			continue
+		}
+
+		var regionsSupportedInstanceType config.InstanceTypeMap
+		provider, ok := m.clusterProvidersConfig.ProvidersConfig.SupportedProviders.GetByName(suCount.CloudProvider)
```

Comment:
I would like to think a little bit about this:

Seems unlikely a cluster will be in a non supported cloudprovider and region, but if that's the case how should we proceed? I think ignoring it at this level would be a good idea. Automatically removing it might be too aggressive/surprising behavior

Reply:
I think it is fine as long as as the cluster is empty, it can be removed. 
There are no region considerations that we can take on the cluster since the region is not supported anymore.
If the region is added back, then we should be fine as dynamic scale up will be bring up the needed cluster.

The approach taken here favors cost saving us leaving the cluster around while it can be deleted is not desirable cost wise, so as long as the cluster (in a region that's does not exists ) does not contain kafka workload we can remove it.